### PR TITLE
Add setting for mandatory Affiliation and Comment fields

### LINKED ISSUE #4819: Mandatory fields "Affiliation" and "Comment" in the account request form
Dear Indico team,

here is a tiny feature request for your consideration.

The problem:

In some cases one needs to enable a registration
for extrnal users. The users, however, too often ignore the fields
"Affiliation" and "Comment" even if they are explicitely asked to fill
them. This creates a situation when the only known information about
the user is the e-mail/name. It hard to decide based on that
information only if one should accept/reject the requst, especially in
the cases when the e-mail adress is from a commercial site.

Possible solution:

Mandaratory fields "Affiliation" and "Comment" in the account request
form would help to get more information for the decision making.

Alternatives:

The solution above + make it possible for indico admins to select if these fields are mandatory

example

Bet regards,

Andrii

### PR DESCRIPTION
**Pull request by elsbethe**
Add a setting for making the 'Affiliation' and 'Comment' field mandatory when creating a new account. When setting is active, the corresponding field is mandatory in the sign up form.

closes #4819

### CODE REVIEW

File: indico/modules/auth/client/js/signup.jsx [hunk 1]
```diff
@@ -141,6 +142,7 @@ function Signup({
               <SyncedFinalInput
                 name="affiliation"
                 label={Translate.string('Affiliation')}
+                required={ moderated && mandatoryFields.includes("Affiliation") }
```
Comment (ThiefMaster) [inline indico/modules/auth/client/js/signup.jsx:145]: please make sure to use eslint so it fixes formatting issues (through its *prettier* integration)

File: indico/modules/auth/client/js/signup.jsx [hunk 2]
```diff
@@ -189,8 +191,9 @@ function Signup({
                 </Translate>
               </Message>
               <FinalTextArea
+                required={ mandatoryFields.includes("Comment") }
```
Comment (ThiefMaster) [inline indico/modules/auth/client/js/signup.jsx:194]: use all lowercase for internal values

File: indico/modules/users/forms.py [hunk 3]
```diff
@@ -116,6 +116,10 @@ class AdminUserSettingsForm(IndicoForm):
                                          description=_('Whether users are allowed to generate personal API tokens. '
                                                        'If disabled, only admins can create them, but users will '
                                                        'still be able to regenerate the tokens assigned to them.'))
+    if config.data['LOCAL_MODERATION']:
+        mandatory_fields_account_request = IndicoSelectMultipleCheckboxField(_('Mandatory fields in account request'),
+                                                                         choices=['Affiliation','Comment'],
```
Comment (ThiefMaster) [inline indico/modules/users/forms.py:121]: I'm pretty sure you can pass a list of tuples here to separate value/title (value should be lowercase, title should be translatable)

**Review by ThiefMaster — COMMENTED** [review]
Thanks! Please also add server-side validation to make sure the mandatory fields are not empty.

**Comment by elsbethe** [comment]
Thanks for your quick feedback! I added the server-side validation and processed your feedback.

File: indico/modules/users/forms.py [hunk 4]
```diff
@@ -116,6 +116,10 @@ class AdminUserSettingsForm(IndicoForm):
                                          description=_('Whether users are allowed to generate personal API tokens. '
                                                        'If disabled, only admins can create them, but users will '
                                                        'still be able to regenerate the tokens assigned to them.'))
+    if config.data['LOCAL_MODERATION']:
```
Comment (ThiefMaster) [inline indico/modules/users/forms.py:119]: - `config.LOCAL_MODERATION`
- the indentation/spacing below is incorrect

File: indico/modules/auth/controllers.py [hunk 5]
```diff
@@ -435,11 +435,17 @@ class Meta:
             first_name = fields.String(required=True)
             last_name = fields.String(required=True)
             address = fields.String(load_default='')
-            affiliation = fields.String(load_default='')
+            if 'affiliation' in user_management_settings.get('mandatory_fields_account_request'):
```
Comment (ThiefMaster) [inline indico/modules/auth/controllers.py:438]: Only apply this if `self.moderate_registrations` is true

File: indico/modules/auth/controllers.py [hunk 6]
```diff
@@ -435,11 +435,17 @@ class Meta:
...
+                affiliation = fields.String(required=True)
```
Comment (ThiefMaster) [inline indico/modules/auth/controllers.py:439]: `validate=not_empty` would be a good idea here as well (otherwise an empty string would be accepted)

**Review by ThiefMaster — APPROVED** [review]

**Pull request merged**